### ISSUE TITLE
Dask Yarn - FileNotFoundError: Failed to resolve .skein.{crt,pem} in 'LOCAL_DIRS'

### ISSUE THREAD

**Issue by yogeshgupta-or** [dask/dask-yarn#56]
I am trying to run my application using dask-yarn but while creating the yarn cluster I am getting an error-

> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/user/project/sync/lib/python3.5/site-packages/dask_yarn/core.py", line 291, in __init__
>     self._start_cluster(spec, skein_client)
>   File "/home/user/project/sync/lib/python3.5/site-packages/dask_yarn/core.py", line 367, in _start_cluster
>     scheduler_address = app.kv.wait('dask.scheduler').decode()
>   File "/home/user/project/sync/lib/python3.5/site-packages/skein/kv.py", line 655, in wait
>     event = event_queue.get()
>   File "/home/user/project/sync/lib/python3.5/site-packages/skein/kv.py", line 281, in get
>     raise out
> skein.exceptions.ConnectionError: Unable to connect to application

On checking the yarn logs I found that some of the worker nodes are throwing FileNotFoundError for skein.{crt,pem}-

> 
> LogContents:
> /home/user/project/sync/lib/python3.5/site-packages/dask_yarn/config.py:13: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
>   defaults = yaml.load(f)
> Traceback (most recent call last):
>   File "/home/user/project/sync/bin/dask-yarn", line 10, in <module>
>     sys.exit(main())
>   File "/home/user/projectn/sync/lib/python3.5/site-packages/dask_yarn/cli.py", line 407, in main
>     func(**kwargs)
>   File "/home/user/project/sync/lib/python3.5/site-packages/dask_yarn/cli.py", line 288, in scheduler
>     app_client = skein.ApplicationClient.from_current()
>   File "/home/user/project/sync/lib/python3.5/site-packages/skein/core.py", line 792, in from_current
>     security=Security.from_default())
>   File "/home/user/project/sync/lib/python3.5/site-packages/skein/model.py", line 344, in from_default
>     "Failed to resolve .skein.{crt,pem} in 'LOCAL_DIRS'")
> FileNotFoundError: Failed to resolve .skein.{crt,pem} in 'LOCAL_DIRS'

Also sometimes my dask job works just fine and sometimes it throws this error. Is this a known issue or am I doing something wrong.

**Comment by jcrist** [comment]
This may be a bug in skein (the YARN python client library backing dask-yarn). YARN is poorly documented, so a lot of its behaviors are reverse engineered, and you may have come across a YARN configuration we haven't seen before.

A few things that would help debug:

- What version of hadoop are you running?
- What is the value of `yarn.nodemanager.local-dirs` in your `yarn-site.xml`? (often in `/etc/hadoop/conf/`, but this is cluster dependent).

**Comment by yogeshgupta-or** [comment]
I am running Hadoop 3. Also since we are using kerberos I am doing kinit before submitting the job. 
My value for  `yarn.nodemanager.local-dirs` is `/mnt/data/bay2/hadoop/yarn/local,/mnt/data/bay4/hadoop/yarn/local`.

Again for some workers files are present in the 'LOCAL-DIRS' but for some it just fails.

**Comment by jcrist** [comment]
Hmmm, ok.

A bit of background here: Skein relies on a few files to being found in each container to run properly. These are transported to each container during YARN's resource localization process, something which should be transparent to the user. During this process one of the directories in `yarn.nodemanager.local-dirs` is chosen to contain a subdirectory where the files will be placed. When the container starts up, this directory will be the process's working directory and contains all of the required resources for it to run.

You've run into an issue that we haven't seen before. Either:
- The required files aren't being properly distributed to the containers
- Or they are, but our code to find them is broken in some way 

I suspect the reason you're seeing some runs work and some fail has to do with the randomness of which directory is chosen.

Could you run the following script please? It will launch a YARN job, log some things, and report back with the application logs. I'd like you to run this until you see `RUN FAILED, PLEASE SUBMIT LOGS` in the output. This will hopefully hit the same issue you were running into and log enough information that we can figure out why. If you find it's not triggering the issue after a reasonable number of runs report back and we can try something else.

*Before posting the output of this script, I recommend checking for any potential things you wouldn't want to share on github. I'm not intentionally logging things that are private, but some file paths, etc... may be specific enough to your job to warrant anonymization*

```python
import subprocess
import time
import skein


def get_logs(appid):
    command = ["yarn", "logs", "-applicationId", appid]
    for i in range(2):
        try:
            return subprocess.check_output(command).decode()
        except Exception:
            time.sleep(0.5)
    return subprocess.check_output(command).decode()


script = """
python -c "
import os

print('CWD: %r' % os.getcwd())
local_dirs = os.environ['LOCAL_DIRS']
print('LOCAL_DIRS: %r' % local_dirs)
container_id = os.environ['CONTAINER_ID']
print('CONTAINER_ID: %r' % container_id)

successful = False
for path in local_dirs.split(','):
    check_dir = os.path.join(path, container_id)
    if os.path.exists(check_dir):
        print('Directory %r exists' % check_dir)
        cert_path = os.path.join(check_dir, '.skein.crt')
        key_path = os.path.join(check_dir, '.skein.pem')
        cert_exists = os.path.exists(cert_path)
        key_exists = os.path.exists(key_path)
        print('Does cert_path %r exist?: %r' % (cert_path, cert_exists))
        print('Does key_path %r exist?: %r' % (key_path, key_exists))
        if cert_exists and key_exists:
            successful = True
        else:
            try:
                print('local directory contents:')
                print(os.listdir(check_dir))
            except Exception as exc:
                print(exc)
    else:
        print('Directory %r does not exist' % check_dir)

if successful:
    print('============================')
    print('RUN PASSED, PLEASE RUN AGAIN')
    print('============================')
else:
    print('==============================')
    print('RUN FAILED, PLEASE SUBMIT LOGS')
    print('==============================')
"
"""

spec = skein.ApplicationSpec(
    name="debug-skein",
    master=skein.Master(
        resources=skein.Resources(
            memory=512,
            vcores=1
        ),
        script=script
    )
)

with skein.Client(log_level='debug') as client:
    appid = client.submit(spec)

    while True:
        state = client.application_report(appid).state
        if state in {'FINISHED', 'FAILED', 'KILLED'}:
            break
        time.sleep(2)

    print("Application %r completed with state %r" % (appid, state))
    print("Getting logs...")
    logs = get_logs(appid)
    print(logs)
```

A good way to run this and log the output to a file would be:

```bash
$  python debug.py 2>&1 | tee log
```

**Comment by yogeshgupta-or** [comment]
I ran the job several times and it passed every time. Here is the log snippet which I feel was relevant for this check-

```
***************************************************************************************

Container: container_e369_1552620760504_0186_01_000001 on some.worker_45454
LogAggregationType: AGGREGATED
============================================================================================
LogType:application.driver.log
LogLastModifiedTime:Mon Apr 01 15:38:46 -0400 2019
LogLength:1517
LogContents:
CWD: '/mnt/data/bay4/hadoop/yarn/local/usercache/user/appcache/application_1552620760504_0186/container_e369_1552620760504_0186_01_000001'
LOCAL_DIRS: '/mnt/data/bay2/hadoop/yarn/local/usercache/user/appcache/application_1552620760504_0186,/mnt/data/bay4/hadoop/yarn/local/usercache/user/appcache/application_1552620760504_0186'
CONTAINER_ID: 'container_e369_1552620760504_0186_01_000001'
Directory '/mnt/data/bay2/hadoop/yarn/local/usercache/user/appcache/application_1552620760504_0186/container_e369_1552620760504_0186_01_000001' exists
Does cert_path '/mnt/data/bay2/hadoop/yarn/local/usercache/user/appcache/application_1552620760504_0186/container_e369_1552620760504_0186_01_000001/.skein.crt' exist?: False
Does key_path '/mnt/data/bay2/hadoop/yarn/local/usercache/user/appcache/application_1552620760504_0186/container_e369_1552620760504_0186_01_000001/.skein.pem' exist?: False
local directory contents:
[]
Directory '/mnt/data/bay4/hadoop/yarn/local/usercache/user/appcache/application_1552620760504_0186/container_e369_1552620760504_0186_01_000001' exists
Does cert_path '/mnt/data/bay4/hadoop/yarn/local/usercache/user/appcache/application_1552620760504_0186/container_e369_1552620760504_0186_01_000001/.skein.crt' exist?: True
Does key_path '/mnt/data/bay4/hadoop/yarn/local/usercache/user/appcache/application_1552620760504_0186/container_e369_1552620760504_0186_01_000001/.skein.pem' exist?: True
============================
RUN PASSED, PLEASE RUN AGAIN
============================

End of LogType:application.driver.log

```

**Comment by jcrist** [comment]
Hmmm, interesting. Looks like all possible application directories are created, but only one is used. I don't know why YARN would do that, but that's a bug in skein, easy to fix. I assume you're using skein 0.6.0 (`skein --version`). If you were to downgrade skein to 0.5.1 things would likely work for you (a change was made recently based on an apparently incorrect assumption about YARN's inner workings).

**Comment by jcrist** [comment]
Should be fixed by https://github.com/jcrist/skein/pull/151. If you want to test, you should be able to build skein locally with `pip install git+https://github.com/jcrist/skein.git@container-dir-more-robust` (requires maven). You could also test by downgrading skein to 0.5.1 - if my diagnosis of the bug was correct the issue was introduced in 0.6.0.

**Comment by yogeshgupta-or** [comment]
Downgrading skein to` 0.5.1` seems to work.

**Comment by jcrist** [comment]
Cool, thanks for working through this issue. I'll push a new release of Skein either today or tomorrow with this patch.

**Comment by jcrist** [comment]
Skein version 0.6.1 (pypi, conda-forge) has been released, as well as dask-yarn 0.5.2 (pypi, still building on conda-forge). Updating should fix this.

Please let me know if you run into any other issues.

**Comment by yogeshgupta-or** [comment]
Thanks Jim, for the fix.
